feat: thea pallet and improvements to asset-handler pallet

felixfaisal opened the pull request:
## Describe your changes

## Issue ticket number and link

## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] I removed all Clippy and Formatting Warnings. 
- [x] I added required Copyrights.

felixfaisal committed:
feat: new extrinsic create_thea_asset for asset_handler

felixfaisal committed:
fix: add tests for creating thea assets

felixfaisal committed:
fix: add tests for bad origins

felixfaisal committed:
fix: making use of identifier length

felixfaisal committed:
fix: add mint and burn functions

felixfaisal committed:
feat: add barebones thea pallet to runtime

felixfaisal committed:
fix: add storage items for approve_deposit

felixfaisal committed:
fix: build errors

felixfaisal committed:
fix: changes to complete approve_deposit

felixfaisal committed:
feat: Add host functions

felixfaisal committed:
feat: Add host functions implementation

felixfaisal committed:
fix: changes to include host functions

felixfaisal committed:
fix: add host functions to pallet and runtime

felixfaisal commented:
Hey @serhii-temchenko The latest CI build has an error related to dependencies 
```bash
error[E0658]: use of unstable library feature 'scoped_threads'
``` 
Can you take a look into this?

felixfaisal committed:
fix: add mock runtime environment file for testing

felixfaisal committed:
fix: tests for approve deposit extrinsic

serhii-temchenko commented:
> Hey @serhii-temchenko The latest CI build has an error related to dependencies
> 
> ```shell
> error[E0658]: use of unstable library feature 'scoped_threads'
> ```
> 
> Can you take a look into this?

Sure, error appeared in the process of formatter installation. Required version (1.5) was outdated and for now in latest rustup installer present 1.5.1 version. So we don't need install formatter separately. I raised the PR #598 with fix of this and some refactoring of workflows declaration. Please review that PR and after merge it - you could pick fix from the `Develop` branch.

felixfaisal committed:
fix: change bitmap from vec to u128

felixfaisal committed:
Merge remote-tracking branch 'origin/Develop' into fx-asset-handler

felixfaisal committed:
fix: apply rustfmt

felixfaisal committed:
fix: apply rustfmt fix

serhii-temchenko commented on pallets/asset-handler/src/lib.rs:
```diff
@@ -298,6 +311,51 @@ pub mod pallet {
+		///
+		/// * `origin`: `Asset` owner
+		/// * `network_id`: Network ID of asset being Bridges
+		/// * `identifier length`: Length of asset identifier length
```
Missed underscore in `identifier length`

felixfaisal replied:
Fixed ✅

felixfaisal committed:
fix: typo for identifier_length

Gauthamastro commented on pallets/asset-handler/src/lib.rs:
```diff
@@ -168,6 +169,12 @@ pub mod pallet {
+	#[pallet::storage]
+	#[pallet::getter(fn get_thea_assets)]
+	pub(super) type TheaAssets<T: Config> =
+		StorageMap<_, Blake2_128Concat, u128, (u8, u8, BoundedVec<u8, ConstU32<1000>>), ValueQuery>;
```
Documentation on this

felixfaisal replied:
Done ✅

Gauthamastro commented on pallets/asset-handler/src/lib.rs:
```diff
@@ -298,6 +311,51 @@ pub mod pallet {
+			let mut derived_asset_id = vec![];
+			derived_asset_id.push(network_id);
+			derived_asset_id.push(identifier_length);
+			derived_asset_id.extend(&asset_identifier[0..identifier_length as usize]);
```
Please add a check to ensure indexing won't panic

felixfaisal replied:
Done ✅

Gauthamastro commented on pallets/asset-handler/src/tests.rs:
```diff
@@ -587,6 +587,64 @@ pub fn test_convert_amount_for_foreign_chain() {
 	});
 }
 
```
Add a case where the recipient account doesn't exist and Thea mints into it. or burns. Let's see the error handling

felixfaisal replied:
Done ✅

felixfaisal committed:
fix: add documentation and handling for index error

felixfaisal committed:
fix: update for existential deposit

serhii-temchenko committed:
Typo fix

serhii-temchenko committed:
Extended and improved thea related tests

serhii-temchenko commented:
@felixfaisal please check PR #602.

serhii-temchenko committed:
Code style fixes

Gauthamastro committed:
Added withdraw ext, on_idle claim of withdrawals, batching of withdrawals

Gauthamastro committed:
Added withdraw ext, on_idle claim of withdrawals, batching of withdrawals

Gauthamastro committed:
Events modified as per frontend requirements

Gauthamastro committed:
batch_withdrawal_complete ext

Gauthamastro committed:
Fix imports

nakul1010 committed:
test cases added

nakul1010 committed:
Fixed clippy warnings and added TC

felixfaisal committed:
fix: clippy warnings

felixfaisal committed:
fix: update tests

felixfaisal committed:
refactor: apply rustfmt

felixfaisal committed:
feat: Thea Pallet (#597)

This PR adds a pallet for Thea and important extrinsics 
- Runtime Inclusion of Pallet 
- Host functions for BLS Functions 
- Mock runtime environment, and tests for bls verification 
- `approve_deposit` extrinsic

felixfaisal committed:
Asset handler tests improvement (#602)

In this PR:
- Common implementation in `thea` related tests moved to separate
private functions to be reused.
- Extended amount of tests to cover more cases
- Added functions with todo's inside for missing cases

serhii-temchenko commented on node/src/service.rs:
```diff
@@ -26,6 +26,7 @@ use polkadex_client::ExecutorDispatch;
 use polkadex_primitives::Block;
 use sc_client_api::{BlockBackend, ExecutorProvider};
 use sc_executor::NativeElseWasmExecutor;
+// use sc_executor::native_executor_instance;
```
Please remove

felixfaisal replied:
Removed ✅

serhii-temchenko commented on node/src/service.rs:
```diff
@@ -44,6 +45,14 @@ type FullSelectChain = sc_consensus::LongestChain<FullBackend, Block>;
 type FullGrandpaBlockImport =
 	sc_finality_grandpa::GrandpaBlockImport<FullBackend, Block, FullClient, FullSelectChain>;
 
+// Our native executor instance.
+// native_executor_instance!(
+// 	pub Executor,
+// 	node_polkadex_runtime::api::dispatch,
+// 	node_polkadex_runtime::native_version,
+// 	// Make the executor aware of the custom host functions.
+// 	thea_primitives::thea_ext::HostFunctions,
+// );
```
Please remove

felixfaisal replied:
Removed ✅

serhii-temchenko commented on pallets/asset-handler/src/lib.rs:
```diff
@@ -221,6 +231,12 @@ pub mod pallet {
 		AllowlistedTokenRemoved,
 		/// Division Overflow
 		DivisionOverflow,
+		// Amount for minting or burning cannot be Zero
+		AmountCannotBeZero,
+		// Thea Asset has not been registered
+		AssetNotRegistered,
+		// Identifier length provided is wrong
```
Please use 3 slashes as above.

felixfaisal replied:
Added ✅

serhii-temchenko commented on pallets/thea/Cargo.toml:
```diff
@@ -0,0 +1,71 @@
+    "pallet-assets/runtime-benchmarks",
+    "chainbridge/runtime-benchmarks",
+    "asset-handler/runtime-benchmarks",
+]
+
+
```
Please remove extra empty line

felixfaisal replied:
Remove ✅

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	// /// Set ID of the current active relayer set
+	// #[pallet::storage]
+	// #[pallet::getter(fn get_current_active_relayer_set_id)]
+	// pub(super) type CurrentActiveRelayerSetId<T: Config> =
+	// StorageMap<_, Blake2_128Concat, u8, u32, OptionQuery>;
```
Please remove if not needed

felixfaisal replied:
Remove ✅

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+	pub(super) type AssetIdToNetworkMapping<T: Config> =
+		StorageMap<_, Blake2_128Concat, u128, Network, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_deposit_nonce)]
+	pub(super) type DepositNonce<T: Config> =
+		StorageMap<_, Blake2_128Concat, Network, u32, ValueQuery>;
```
Please add documentation as well

felixfaisal replied:
Added ✅

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+							// Force push is fine as it was part of the bounded vec
+							pending_deposits.force_push(deposit);
+							// We can't do much here other than to log an error.
+							log::error!(target:"runtime::thea::on_idle","Error while claiming deposit on idle: user: {:?}, Err: {:?}",account,err);
+						}
+						// reduce the remaining_weight
+						remaining_weight = remaining_weight.saturating_sub(single_claim_weight);
+						if remaining_weight.is_zero() {
+							break
+						}
+					}
+
+					if !pending_deposits.is_empty() {
+						<ApprovedDeposits<T>>::insert(&account, pending_deposits);
+						accounts.insert(account);
+					}
+				}
+			}
+
+			<AccountWithPendingDeposits<T>>::put(accounts);
```
Is it ok that we are popping account on line 227 and if no approved deposits for this account yet - we just removing it from storage by overwriting here on line 251? shouldn't we put it back in accounts collection if there's no approved deposits yet? if no - at least some kind of log should be in `else` case between 248 and 249 lines.

felixfaisal replied:
It's fine, All the accounts are accounts that have pending deposits, which is being called here 
```rust
let mut accounts = <AccountWithPendingDeposits<T>>::get();
```
CC: @Gauthamastro

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+		/// * `amount`: The amount of assets that have been deposited in foreign chain
+		/// * `bls_signature`: The aggregated signature of majority of relayers in current active
+		///   relayer set
+		#[pallet::weight(1000)]
```
Is there a task for the benchmark?

felixfaisal replied:
Yes, I've created an issue for it

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+		///   relayer set
+		#[pallet::weight(1000)]
+		pub fn approve_deposit(
+			origin: OriginFor<T>,
+			bit_map: u128,
+			bls_signature: [u8; 96],
+			payload: Payload<T::AccountId>,
```
Dockblock has a different set of args. Please update.

felixfaisal replied:
Updated ✅

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			ensure!(payload.amount > 0, Error::<T>::AmountCannotBeZero);
+			// Fetch Deposit Nonce
+			let nonce = <DepositNonce<T>>::get(payload.network_id.saturated_into::<Network>());
+			ensure!(payload.deposit_nonce == nonce + 1, Error::<T>::DepositNonceError);
```
I suggest to move this bloc after line 303 since its required on line 306 only.

felixfaisal replied:
I don't get this comment, can you please elaborate

serhii-temchenko replied:
Sure, this selected block could be moved lower by execution chain since `nonce` required only on line number 306. So by moving 281-283 lines after 303 - other assertions would be executed first and calculation time in invalid scenario will be faster.

felixfaisal replied:
Other assertions are computationally more expensive, So isn't it better to have it here?

serhii-temchenko replied:
Ok

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+
+			// Fetch current active relayer set BLS Keys
+			let current_active_relayer_set =
+				Self::get_relayers_key_vector(payload.network_id).unwrap();
```
Unhandled unwrap.

felixfaisal replied:
Removed unwrap ✅

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+			if <ApprovedDeposits<T>>::contains_key(&payload.who) {
+				<ApprovedDeposits<T>>::mutate(payload.who.clone(), |bounded_vec| {
+					if let Some(inner_bounded_vec) = bounded_vec {
+						inner_bounded_vec.try_push(approved_deposit).unwrap();
```
Unhandled unwrap

felixfaisal replied:
This will be refactored, Will add a comment

serhii-temchenko replied:
Still unresolved

felixfaisal replied:
Added a comment

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+			};
+			if <ApprovedDeposits<T>>::contains_key(&payload.who) {
+				<ApprovedDeposits<T>>::mutate(payload.who.clone(), |bounded_vec| {
+					if let Some(inner_bounded_vec) = bounded_vec {
```
If it's not - shouldn't be created vector and still pushed approved deposit?

felixfaisal replied:
The scenario is not possible, if `Storage item` contains the key

serhii-temchenko commented on thea-primitives/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+sp-runtime-interface = { git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.28",default-features = false }
```
Please add space after comma `"polkadot-v0.9.28",default-features`

felixfaisal replied:
Updated ✅

serhii-temchenko commented on thea-primitives/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+    "frame-system/std",
+    "parity-scale-codec/std",
+    "scale-info/std",
+]
```
Please add empty line

felixfaisal replied:
Added ✅

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+		pub fn claim_deposit(origin: OriginFor<T>, num_deposits: u32) -> DispatchResult {
+			let user = ensure_signed(origin)?;
+
+			if let Some(mut deposits) = <ApprovedDeposits<T>>::get(&user) {
```
Is it proper logic to do nothing, when calling this extrinsic, if no approved deposits? mby some proper error code should be returned of at least logged? or it should be idempotent?

felixfaisal replied:
Added an else case, which returns an error

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+					// If pending deposits are available, save it back
+					<ApprovedDeposits<T>>::insert(&user, deposits)
+				} else {
+					<AccountWithPendingDeposits<T>>::mutate(|accounts| accounts.remove(&user));
```
Is it possible that by time of execution of loop from above some new elements could appeared in storage (ApprovedDeposits)? so we will end up with hanging ApprovedDeposits since account was removed from the pending for processing list?

felixfaisal replied:
This race condition should not be possible, It will not be executed in parallel, The `on_idle` hook will only make use of any remaining weight if available, that is when there is no activity in the blockchain. 
CC: @zktony @Gauthamastro

Gauthamastro replied:
Yes, on_idle will execute only when there are no more extrinsic to include in the block and enough remaining weight to complete the approve_deposit.

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+				.saturated_into();
+
+				for _ in 0..length {
+					if let Some(deposit) = deposits.pop() {
+						if let Err(err) = Self::execute_deposit(deposit, &user) {
+							// Force push is fine as it will have the capacity.
+							deposits.force_push(deposit);
+							// Save it back on failure
+							<ApprovedDeposits<T>>::insert(&user, deposits.clone());
+							return Err(err)
+						}
+					} else {
+						break
+					}
```
Shouldn't this deposit (if no errors on calling `execute_deposit`) be removed from the storage (ApprovedDeposits) since transaction was made to not make it permanent?

felixfaisal replied:
We are beginning this code block with `deposits.pop()` so if it does not return any error, It's not in the storage item, and then after the loop we are overwriting the `Storage Item`, `Approved Deposits` with the `deposits` vector. So it is being removed.

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			// TODO: Something to verify BLS signature
```
When this should be implemented? please add in-code reference to the issue

felixfaisal replied:
This will be implemented when work on withdrawal begins

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+		/// * `amount`: Amount of asset to withdraw
+		#[pallet::weight(1000)]
+		pub fn withdraw(
+			origin: OriginFor<T>,
+			asset_id: u128,
+			amount: u128,
+			beneficiary: Vec<u8>,
+			pay_for_remaining: bool,
```
Dockblock has different amount of parameters. Please update.

felixfaisal replied:
Updated ✅

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+
+			// Find native network of this asset
+			#[allow(clippy::unnecessary_lazy_evaluations)]
+			// TODO: Remove once merged to asset-handler
```
Please elaborate on this comment

felixfaisal replied:
This will be refactored when work on withdrawal so not fixing clippy suggestion for now

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+			ensure!(pending_withdrawals.is_full(), Error::<T>::WithdrawalNotAllowed);
+
+			#[allow(clippy::unnecessary_lazy_evaluations)]
+			// TODO: Remove once merged to asset-handler
```
Same

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -0,0 +1,534 @@
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			// TODO: Update Thea Staking pallet about fees collected
```
Please add some task reference to not forget

felixfaisal replied:
Done

serhii-temchenko reviewed (changes requested):
Only 3 assertions present in a tests for `approve_deposit` extrinsic. Please add all required tests and assertions for the `thea` pallet.

felixfaisal committed:
feat: remove Payload from thea-pallet code

felixfaisal committed:
refactor: apply review comments

felixfaisal committed:
refactor: update doc

felixfaisal committed:
fix: remove unhandled unwrap

felixfaisal committed:
refactor: review comments

felixfaisal committed:
refactor: rustfmt

nakul1010 committed:
Merge branch 'Develop' of https://github.com/Polkadex-Substrate/Polkadex into fx-asset-handler

# Conflicts:
#	Cargo.lock
#	Cargo.toml
#	runtime/Cargo.toml
#	runtime/src/lib.rs

nakul1010 committed:
handled conflicts

felixfaisal committed:
fix: failing test

felixfaisal committed:
Fix PR review comments  (#613)

This PR fixes all review comments in the above branch.

serhii-temchenko committed:
Added todos for benchmarking

serhii-temchenko commented:
@felixfaisal still 2 comments untouched. Please check them. 
Plus nothing about [this](https://github.com/Polkadex-Substrate/Polkadex/pull/595#pullrequestreview-1246288395) comment.

felixfaisal commented:
Hey @serhii-temchenko Tests will be covered under a separate issue - https://github.com/Polkadex-Substrate/Polkadex/issues/614

felixfaisal committed:
refactor: add comment
